### PR TITLE
Store port values as u16 in Member struct

### PR DESCRIPTION
**Pull request by baumanj**
This is more consistent with the rest of the local representations. Since
protobuf doesn't support 16-bit ints, we have to convert on that boundary, but
we should be checking values there and avoiding unsafe casting elsewhere.

AtomicUsize instances would be converted to AtomicU16s, but integer_atomics
isn't stable yet, so we use Mutex<u16>.
See https://github.com/rust-lang/rust/issues/32976

Signed-off-by: Jon Bauman <5906042+baumanj@users.noreply.github.com>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/butterfly/src/member.rs [hunk 1]
```diff
@@ -204,8 +204,8 @@ impl From<Member> for proto::Member {
             id: Some(value.id),
             incarnation: Some(value.incarnation.to_u64()),
             address: Some(value.address),
-            swim_port: Some(value.swim_port),
-            gossip_port: Some(value.gossip_port),
+            swim_port: Some(value.swim_port.into()),
+            gossip_port: Some(value.gossip_port.into()),
```
Comment (baumanj) [inline components/butterfly/src/member.rs]: u16 → i32 can be done safely

File: components/butterfly-test/src/lib.rs [hunk 2]
```diff
@@ -53,14 +57,20 @@ impl Suitability for NSuitability {
...
-    let gossip_port = SERVER_PORT.fetch_add(1, Ordering::Relaxed);
+    let swim_port;
+    let gossip_port;
+    {
+        let mut port_guard = SERVER_PORT.lock().expect("SERVER_PORT mutex poisoned");
+        swim_port = *port_guard;
+        gossip_port = *port_guard;
+        *port_guard += 1;
+        *port_guard += 1;
+    }
```
[the remark this answers is not in the document]
Comment (baumanj) [reply components/butterfly-test/src/lib.rs]: Good catch! You're right, it should be:
```rust
        swim_port = *port_guard;
        *port_guard += 1;
        gossip_port = *port_guard;
        *port_guard += 1;
```

And that makes me wonder, does this code not get exercised at all in our automation, or is this broken in some other way that the failure isn't reported? I'll investigate.

File: components/butterfly/src/member.rs [hunk 3]
```diff
@@ -229,6 +229,18 @@ impl From<Membership> for proto::Membership {
...
 }
 
+/// Since protobuf doesn't have support for 16-bit ints, we need to check that
+/// we haven't received something illegal
+fn as_port(x: i32) -> Option<u16> {
+    const PORT_MIN: i32 = ::std::u16::MIN as i32;
+    const PORT_MAX: i32 = ::std::u16::MAX as i32;
+
+    match x {
+        PORT_MIN..=PORT_MAX => Some(x as u16),
```
[the remark this answers is not in the document]
Comment (raskchanky) [reply components/butterfly/src/member.rs]: Looks like `..=` is new in 2018 Rust.

https://doc.rust-lang.org/stable/book/2018-edition/appendix-02-operators.html

[the remark this answers is not in the document]
Comment (baumanj) [reply components/butterfly/src/member.rs]: I could never remember which of `...` and `..` were inclusive or exclusive, so I'm a big fan of `..=`.

File: components/butterfly-test/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (baumanj) [reply components/butterfly-test/src/lib.rs]: The [butterfly integration tests](https://github.com/habitat-sh/habitat/blob/master/components/butterfly/tests/integration.rs#L28) do use this code, and you can see (e.g., `two_members_meshed_confirm_one_member` them running successfully here): https://travis-ci.org/habitat-sh/habitat/jobs/443363704.

I'm confused how that's passing. I'll dig deeper.

[the remark this answers is not in the document]
Comment (baumanj) [reply components/butterfly-test/src/lib.rs]: Apparently in normal operation [`swim_addr` and `gossip_addr` are the same](https://github.com/habitat-sh/habitat/blob/master/components/sup/src/manager/mod.rs#L363-L364). And it appears there's no UX for ever making them different. So, I wonder why we're bothering to have different addresses used in the tests. Possibly, we should eliminate [the separate arguments in `habitat_butterfly::server::Server::new`](https://github.com/habitat-sh/habitat/blob/master/components/butterfly/src/server/mod.rs#L276-L277), but for now, I'll just make the fix to maintain the existing behavior.

**Pull request merged**